### PR TITLE
Create AppsV1Deployment resource lifecycle test - +6 endpoint coverage

### LINKED ISSUE #90916: Write AppsV1Deployment resource lifecycle test - +6 endpoint coverage
# Progress `[6/7]`
* [x] Original approval issue: #89340
* [x]  Test approval issue: [kubernetes/kubernetes#90916](https://github.com/kubernetes/kubernetes/issues/90916)
* [x]  Test pr: #92589
* [x]  Two weeks soak start date: 5 Nov 2020 [testgrid-link](https://testgrid.k8s.io/sig-apps#gce&include-filter-by-regex=Deployment&include-filter-by-regex=should%20run%20the%20lifecycle%20of%20a%20Deployment)
* [x]  Code update to remove single flake PR [#96384](https://github.com/kubernetes/kubernetes/pull/96384)
* [x]  Two weeks soak end date: 19 November 2020
* [x]  Test promotion pr: [96487](https://github.com/kubernetes/kubernetes/pull/96487)

This issue is created to allow edit by @Riaankl

/wip
/hold

# Identifying an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining Deployment endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM untested_stable_endpoints
  where kind like 'Deployment'
  -- WHERE operation_id ilike '%Deployment%'
 ORDER BY kind,operation_id desc
 LIMIT 25
       ;
```

```
                operation_id                |    kind    
--------------------------------------------+------------
 replaceAppsV1NamespacedDeploymentStatus    | Deployment
 readAppsV1NamespacedDeploymentStatus       | Deployment
 patchAppsV1NamespacedDeploymentStatus      | Deployment
 patchAppsV1NamespacedDeployment            | Deployment
 listAppsV1DeploymentForAllNamespaces       | Deployment
 deleteAppsV1CollectionNamespacedDeployment | Deployment
(6 rows)
```

# API Reference and feature documentation
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - Deployment](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/Deployment.go)

# The mock test
## Test outline
1. Create a Deployment with a static label
2. Patch the Deployment with a new Label and updated data
3. Get the Deployment to ensure it's patched
4. List all Deployments in all Namespaces find the Deployment(1) ensure that the Deployment is found and is patched
5. Delete Namespaced Deployment(1) via a Collection with a LabelSelector

## Test the functionality in Go
```go
package main

import (
  "encoding/json"
  "fmt"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  appsv1 "k8s.io/api/apps/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/apimachinery/pkg/runtime"
  "k8s.io/apimachinery/pkg/runtime/schema"
  //"k8s.io/apimachinery/pkg/conversion/unstructured"
  unstructuredv1 "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
  watch "k8s.io/apimachinery/pkg/watch"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
      fmt.Println(err)
      return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  deploymentResource := schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments"}

  testDeploymentName := "test-deployment"
  testDeploymentInitialImage := "nginx"
  testDeploymentPatchImage := "alpine"
  testDeploymentUpdateImage := "httpd"
  testDeploymentDefaultReplicas := int32(3)
  testDeploymentMinimumReplicas := int32(1)
  testDeploymentNoReplicas := int32(0)
  testDeploymentLabelSelectors := metav1.LabelSelector{
      MatchLabels: map[string]string{"app": "test-deployment"},
  }
  testNamespaceName := "default"

  fmt.Println("creating a Deployment")
  testDeployment := appsv1.Deployment{
      ObjectMeta: metav1.ObjectMeta{
          Name: testDeploymentName,
          Labels: map[string]string{"test-deployment-static": "true"},
      },
      Spec: appsv1.DeploymentSpec{
          Replicas: &testDeploymentDefaultReplicas,
          Selector: &testDeploymentLabelSelectors,
          Template: v1.PodTemplateSpec{
              ObjectMeta: metav1.ObjectMeta{
                  Labels: testDeploymentLabelSelectors.MatchLabels,
              },
              Spec: v1.PodSpec{
                  Containers: []v1.Container{{
                      Name: testDeploymentName,
                      Image: testDeploymentInitialImage,
                  }},
              },
          },
      },
  }
  _, err = ClientSet.AppsV1().Deployments(testNamespaceName).Create(&testDeployment)
  if err != nil {
      fmt.Println(err)
      return
  }

  fmt.Println("watching for the Deployment to be added")
  dplmtWatchTimeoutSeconds := int64(180)
  dplmtWatch, err := ClientSet.AppsV1().Deployments(testNamespaceName).Watch(metav1.ListOptions{LabelSelector: "test-deployment-static=true", TimeoutSeconds: &dplmtWatchTimeoutSeconds})
  if err != nil {
      fmt.Println(err, "Failed to setup watch on newly created Deployment")
      return
  }

  dplmtWatchChan := dplmtWatch.ResultChan()
  for event := range dplmtWatchChan {
      if event.Type == watch.Added {
          break
      }
  }
  defer func() {
    fmt.Println("deleting the Deployment")
    err = ClientSet.AppsV1().Deployments(testNamespaceName).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-deployment-static=true"})
    if err != nil {
      fmt.Println(err)
      return
    }
    for event := range dplmtWatchChan {
      deployment, ok := event.Object.(*appsv1.Deployment)
      if ok != true {
        fmt.Println("unable to convert event.Object type")
        return
      }
      if event.Type == watch.Deleted && deployment.ObjectMeta.Name == testDeploymentName {
        break
      }
    }
  }()
  fmt.Println("waiting for all Replicas to be Ready")
  for event := range dplmtWatchChan {
      deployment, ok := event.Object.(*appsv1.Deployment)
      if ok != true {
          fmt.Println("unable to convert event.Object type")
          return
      }
      if deployment.Status.AvailableReplicas == testDeploymentDefaultReplicas &&
         deployment.Status.ReadyReplicas == testDeploymentDefaultReplicas {
          break
      }
  }

  fmt.Println("patching the Deployment")
  deploymentPatch, err := json.Marshal(map[string]interface{}{
      "metadata": map[string]interface{}{
          "labels": map[string]string{"test-deployment": "patched"},
      },
      "spec": map[string]interface{}{
          "replicas": testDeploymentMinimumReplicas,
          "template": map[string]interface{}{
              "spec": map[string]interface{}{
                  "containers": []map[string]interface{}{{
                      "name": testDeploymentName,
                      "image": testDeploymentPatchImage,
                      "command": []string{"/bin/sleep", "100000"},
                  }},
              },
          },
      },
  })
  if err != nil {
      fmt.Println(err, "failed to Marshal Deployment JSON patch")
      return
  }
  _, err = ClientSet.AppsV1().Deployments(testNamespaceName).Patch(testDeploymentName, types.StrategicMergePatchType, []byte(deploymentPatch))
  if err != nil {
       fmt.Println(err, "failed to patch Deployment")
       return
  }

  for event := range dplmtWatchChan {
      if event.Type == watch.Modified {
          break
      }
  }
  fmt.Println("waiting for Replicas to scale")
  for event := range dplmtWatchChan {
      deployment, ok := event.Object.(*appsv1.Deployment)
      if ok != true {
          fmt.Println("unable to convert event.Object type")
          return
      }
      if deployment.Status.AvailableReplicas == testDeploymentMinimumReplicas &&
         deployment.Status.ReadyReplicas == testDeploymentMinimumReplicas {
          break
      }
  }


  fmt.Println("listing Deployments")
  deploymentsList, err := ClientSet.AppsV1().Deployments("").List(metav1.ListOptions{LabelSelector: "test-deployment-static=true"})
  if err != nil {
      fmt.Println(err, "failed to list Deployments")
      return
  }
  foundDeployment := false
  for _, deploymentItem := range deploymentsList.Items {
      if deploymentItem.ObjectMeta.Name == testDeploymentName &&
         deploymentItem.ObjectMeta.Namespace == testNamespaceName &&
         deploymentItem.ObjectMeta.Labels["test-deployment-static"] == "true" &&
         *deploymentItem.Spec.Replicas == testDeploymentMinimumReplicas &&
         deploymentItem.Spec.Template.Spec.Containers[0].Image == testDeploymentPatchImage {
          foundDeployment = true
          break
      }
  }
  if foundDeployment != true {
      fmt.Println("unable to find the Deployment in list")
      return
  }

  fmt.Println("updating the DeploymentStatus")
  testDeploymentUpdate := testDeployment
  testDeploymentUpdate.ObjectMeta.Labels["test-deployment"] = "updated"
  testDeploymentUpdate.Spec.Template.Spec.Containers[0].Image = testDeploymentUpdateImage
  testDeploymentDefaultReplicasPointer := &testDeploymentDefaultReplicas
  testDeploymentUpdate.Spec.Replicas = testDeploymentDefaultReplicasPointer
  testDeploymentUpdate.Status.ReadyReplicas = testDeploymentNoReplicas
  testDeploymentUpdateUnstructuredMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&testDeploymentUpdate)
  if err != nil {
      fmt.Println(err, "failed to convert to unstructured")
  }
  testDeploymentUpdateUnstructured := unstructuredv1.Unstructured{
      Object: testDeploymentUpdateUnstructuredMap,
  }
	// currently this hasn't been able to hit the endpoint replaceAppsV1NamespacedDeploymentStatus
  _, err = DynamicClientSet.Resource(deploymentResource).Namespace(testNamespaceName).Update(&testDeploymentUpdateUnstructured, metav1.UpdateOptions{})//, "status")
  if err != nil {
      fmt.Println(err, "failed to update the DeploymentStatus")
      return
  }
  for event := range dplmtWatchChan {
      if event.Type == watch.Modified {
          break
      }
  }

  fmt.Println("fetching the DeploymentStatus")
  deploymentGetUnstructured, err := DynamicClientSet.Resource(deploymentResource).Namespace(testNamespaceName).Get(testDeploymentName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err, "failed to fetch the Deployment")
      return
  }
  deploymentGet := appsv1.Deployment{}
  err = runtime.DefaultUnstructuredConverter.FromUnstructured(deploymentGetUnstructured.Object, &deploymentGet)
  if err != nil {
      fmt.Println(err, "failed to convert the unstructured response to a Deployment")
      return
  }
  if ! (deploymentGet.Spec.Template.Spec.Containers[0].Image == testDeploymentUpdateImage || deploymentGet.Status.ReadyReplicas == testDeploymentNoReplicas || deploymentGet.ObjectMeta.Labels["test-deployment"] == "updated") {
      fmt.Println("failed to update the Deployment (did not return correct values)")
      return
  }
  for event := range dplmtWatchChan {
      if event.Type == watch.Modified {
          break
      }
  }
  for event := range dplmtWatchChan {
      deployment, ok := event.Object.(*appsv1.Deployment)
      if ok != true {
          fmt.Println("failed to convert event Object to a Deployment")
          return
      }
      if deployment.Status.ReadyReplicas == testDeploymentDefaultReplicas {
          break
      }
  }

  fmt.Println("patching the DeploymentStatus")
  deploymentStatusPatch, err := json.Marshal(map[string]interface{}{
      "metadata": map[string]interface{}{
          "labels": map[string]string{"test-deployment": "patched-status"},
      },
      "status": map[string]interface{}{
          "readyReplicas": testDeploymentNoReplicas,
      },
  })
  if err != nil {
      fmt.Println(err, "failed to Marshal Deployment JSON patch")
      return
  }
  DynamicClientSet.Resource(deploymentResource).Namespace(testNamespaceName).Patch(testDeploymentName, types.StrategicMergePatchType, []byte(deploymentStatusPatch), metav1.PatchOptions{}, "status")

  fmt.Println("fetching the DeploymentStatus")
  deploymentGetUnstructured, err = DynamicClientSet.Resource(deploymentResource).Namespace(testNamespaceName).Get(testDeploymentName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err, "failed to fetch the DeploymentStatus")
      return
  }
  deploymentGet = appsv1.Deployment{}
  err = runtime.DefaultUnstructuredConverter.FromUnstructured(deploymentGetUnstructured.Object, &deploymentGet)
  if err != nil {
      fmt.Println(err, "failed to convert the unstructured response to a Deployment")
      return
  }
  if ! (deploymentGet.Spec.Template.Spec.Containers[0].Image == testDeploymentUpdateImage || deploymentGet.Status.ReadyReplicas == 0 || deploymentGet.ObjectMeta.Labels["test-deployment"] == "patched-status") {
      fmt.Println("failed to update the Deployment (did not return correct values)")
      return
  }
  for event := range dplmtWatchChan {
      if event.Type == watch.Modified {
          break
      }
  }
  for event := range dplmtWatchChan {
      deployment, ok := event.Object.(*appsv1.Deployment)
      if ok != true {
          fmt.Println("failed to convert event Object to a Deployment")
          return
      }
      if deployment.Status.ReadyReplicas == testDeploymentDefaultReplicas {
          break
      }
  }

  // write test here
  fmt.Println("[status] complete")
}
```

```go
creating a Deployment
watching for the Deployment to be added
waiting for all Replicas to be Ready
patching the Deployment
waiting for Replicas to scale
listing Deployments
updating the DeploymentStatus
fetching the DeploymentStatus
patching the DeploymentStatus
fetching the DeploymentStatus
[status] complete
deleting the Deployment
```

# Verifying increase it coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |                operation_id                | hit_by_ete | hit_by_new_test 
-------------------+--------------------------------------------+------------+-----------------
 live-test-writing | createAppsV1NamespacedDeployment           | t          |               1
 live-test-writing | deleteAppsV1CollectionNamespacedDeployment | f          |               1
 live-test-writing | listAppsV1DeploymentForAllNamespaces       | f          |               1
 live-test-writing | listAppsV1NamespacedDeployment             | t          |               1
 live-test-writing | patchAppsV1NamespacedDeployment            | f          |               1
 live-test-writing | patchAppsV1NamespacedDeploymentStatus      | f          |               1
 live-test-writing | readAppsV1NamespacedDeploymentStatus       | f          |               2
 live-test-writing | replaceAppsV1NamespacedDeployment          | t          |               1
(8 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------+-----------------+--------------+--------------+------------------
 test_coverage |             445 |          181 |          186 |                5
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 5 points****

This test is also created with the goal of conformance promotion.

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
deleteAppsV1CollectionNamespacedDeployment
listAppsV1DeploymentForAllNamespaces
patchAppsV1NamespacedDeployment
patchAppsV1NamespacedDeploymentStatus
readAppsV1NamespacedDeploymentStatus
replaceAppsV1NamespacedDeploymentStatus

**Which issue(s) this PR fixes:**
Fixes #90916

**Special notes for your reviewer:**
Adds +5 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE
```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE
```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/assign @spiffxp @johnbelamaric 

@BobyMCbobs update this test and removed Watch tooling
Can we please get an /LGTM and /approve

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:286]: Why is `dc.Resource` being used instead of `f.ClientSet.AppsV1().Deployments`?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:212]: Have you considered using watchtools.Until?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:243]: This isn't waiting for scale, the next loop is.  I would remove this, it seems redundant.

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:272]: I might swap this with the loop above.  Verify that patch has taken effect on the deployment spec before waiting for its status to reconcile.

File: test/e2e/apps/deployment.go
Comment (spiffxp) [reply test/e2e/apps/deployment.go:286]: I'm also curious why `Update` is being used here instead of the usual marshal-json-to-bytes-then-`Patch` approach I've seen in other tests from you?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:274]: This isn't actually updating deploymentstatus?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:312]: What are we waiting for here exactly?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:295]: What are we waiting for here exactly?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:318]: I didn't see us patch replicas, what action have we taken that requires us to wait for replicas to reach their default spec?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:359]: Why are we waiting on watch events twice?

File: test/e2e/apps/deployment.go
Comment (spiffxp) [inline test/e2e/apps/deployment.go:374]: Verifying we saw a deleted watch event for a resource is not the same thing, IMO, as verifying that the resource has been deleted (and is no longer retrievable)

**Review by spiffxp — CHANGES_REQUESTED** [review]
This needs work

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Not sure why this landed in this PR, I didn't do that manually.
Reverted in 9c2330db4c2

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go:286]: > Why is `dc.Resource` being used instead of `f.ClientSet.AppsV1().Deployments`?

This was meant to hit _replaceAppsV1NamespacedDeploymentStatus_, which to my research requires the DynamicClient - also why it's still remaining like this. Should I replace it with the default ClientSet, or should I leave it around for when the Endpoint could be added at a later date as a reminder?

Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go:286]: > I'm also curious why `Update` is being used here instead of the usual marshal-json-to-bytes-then-`Patch` approach I've seen in other tests from you?

Also because it's trying to hit the endpoint _replaceAppsV1NamespacedDeploymentStatus_.

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go:212]: Thank you for the suggestion, it's seemed to work really well in other current PRs. I'll adapt it in this test too

Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Updated in fc8e24c6869

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Listing the Deployments in all namespaces here was originally put in here so the endpoint would be hit but also running checks since we have some data. I think it's OK to swap it out as you suggest, being as the same call is made at the start of the test (to get the beginning resource version)

Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: However, it might be good to still have the functionality of still finding the deployment, created in this test, in a list of all deployments in all namespaces.

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go:280]: This is another remainder or the endpoint which was meant to be hit. I'll clean this one up

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: The pattern that this test follows is, to after each modification, it will be verified that it has been modified.
This is useful for making sure there aren't conflicts with things like updates.

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: I believe that I put this in originally, so it would make sure that the replicas are set to the _default_ amount **before* patching, for replica check accuracy reasons

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Doing multiple checks like this seemed to me to be the cleanest way to checking for multiple separate things

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Should this check for something like if the _Meta.DeletionTimestamp_ is not `nil`?

File: test/e2e/apps/deployment.go
Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: The comment has been fixed. I believe it's good to have a confirm on the patch/update when it comes through for data/resource consistency (such as, we have the correct _resourceVersion_, data, etc...)

**Comment by hh** [comment]
@justaugustus @dims @spiffxp can I get a /release 1.19 tag?

File: test/e2e/apps/deployment.go
Comment (dims) [inline test/e2e/apps/deployment.go:149]: @Riaankl @hh if this is going to be a conformance, then we should make sure we use the images from the ones we use for conformance. right?

Comment (BobyMCbobs) [reply test/e2e/apps/deployment.go]: Update to use conformance approved images n ee29022ac8a

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/milestone v1.19
/priority important-soon

**Comment by liggitt** [comment]
Newly introduced test is flaking, see:

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286504904986202114
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286501878984609792
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286495839270735874
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286401342444670980
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286392657853026304
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286377181127118850

revert in https://github.com/kubernetes/kubernetes/pull/93405

**Comment by liggitt** [comment]
it looks like the test added in this PR was the only failure in four consecutive presubmit runs:

* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92589/pull-kubernetes-e2e-kind/1286412665119313926
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92589/pull-kubernetes-e2e-kind/1286401342444670980
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92589/pull-kubernetes-e2e-kind/1286392657853026304
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92589/pull-kubernetes-e2e-kind/1286377181127118850

that should have been a sign that there were issues with the test that needed to be resolved before merging

**Comment by Huang-Wei** [comment]
Can we squash the commits before merging this kind of PR? The unnecessary intermediate commits make the git history hard to browse:

image

**Comment by BobyMCbobs** [comment]
> Can we squash the commits before merging this kind of PR? The unnecessary intermediate commits make the git history hard to browse:
> 
> image

My mistake, that doesn't look great. I'm gonna fix my flow of doing things so this doesn't happen.

**Pull request merged**
